regnety/dataset/imagenet.py

```diff
@@ -0,0 +1,268 @@
+import tensorflow as tf
+import os
+from official.vision.image_classification.augment import RandAugment
+
+
+class ImageNet:
+    """Class for all ImageNet related functions, includes TFRecords loading,
+    preprocessing and augmentations. TFRecords must follow the format given
+    below.
+
+    Args:
+        tfrecs_filepath: list of filepaths of all TFRecords files
+        batch_size: batch_size for the Dataset
+        image_size: final image size of the images in the dataset
+        augment_fn: function to apply to dataset after loading raw TFrecords
+                    'default' implies following augmentations will be applied:
+                    - Random sized crop (train only)
+                    - Scale and center crop
+                    - RandAugment (TODO)
+        num_classes: number of classes
+        randaugment: True of RandAugment is to be applied
+
+    Class attributes:
+        TFRECS_FORMAT: Expected format of TFRecords stored.
+        _MEAN: Channel wise mean for ImageNet
+        _STD: Channel wise standard deviation for ImageNet
+
+    """
+
+    TFRECS_FORMAT = {
+        "image": tf.io.FixedLenFeature([], tf.string),
+        # "height": tf.io.FixedLenFeature([], tf.int64),
+        # "width": tf.io.FixedLenFeature([], tf.int64),
+        "filename": tf.io.FixedLenFeature([], tf.string),
+        "label": tf.io.FixedLenFeature([], tf.int64),
+        "synset": tf.io.FixedLenFeature([], tf.string),
+    }
+
+    _MEAN = tf.constant([0.485, 0.456, 0.406])
+    _STD = tf.constant([0.229, 0.224, 0.225])
+
+    def __init__(
+        self,
+        tfrecs_filepath=None,
+        batch_size=128,
+        image_size=224,
+        augment_fn="default",
+        num_classes=10,
+        randaugment=True,
+    ):
+
+        if tfrecs_filepath is None:
+            raise ValueError("List of TFrecords paths cannot be None")
+        self.tfrecs_filepath = tfrecs_filepath
+        self.batch_size = batch_size
+        self.image_size = image_size
+        self.augment_fn = augment_fn
+        self.num_classes = num_classes
+        self.randaugment = randaugment
+        if self.randaugment:
+            self.augmenter = RandAugment(magnitude=5)
+
+    def decode_example(self, example):
+        """Decodes an example to its individual attributes
+
+        Args:
+            example: A TFRecord dataset example.
+
+        Returns:
+            Dict containing attributes from a single example. Follows
+            the same names as TFRECORDS_FORMAT.
```

Comment:
> It looks like this return data isn't a perfect match for what's inside the TFRecord

I changed the return data format slightly because it was decided to resize the images to 512. Thus, height and width become implicit (for our use case). This is in continuation to line 81 in `tfrecords_utils.py` .

Reply:
>consider defining a simple dataclass specifically for this return value.

@MorganR 
If I understand correctly, you were saying that I should make a dataclass and use it as a common return object for all functions which we map to the dataset (like `decode_example` in this case). In that case, I think tf.data.Dataset allows only tuples,lists or dicts as its elements. Please correct me if I'm wrong.

>This would provide an even better IDE experience if you add type annotations


Sure, I'll add type annotations.

Reply:
Makes sense. I added this comment before reading the rest of the files.